odoo/addons/base/tests/test_ir_cron.py

```diff
@@ -29,40 +37,364 @@ def capture_triggers(self, cron_id=None):
+
+        with self.subTest(doall=True, numbercall=-1, missed_call=2), \
+             patch.object(self.registry['ir.cron'], '_callback') as callback:
+            self.cron.active = True
+            self.cron.doall = True
+            self.cron.numbercall = -1
+            self.cron.nextcall = fields.Datetime.now() - timedelta(days=1)
+            process_job()
+            self.assertEqual(callback.call_count, 2)
+            self.assertEqual(self.cron.lastcall, fields.Datetime.now())
+            self.assertEqual(self.cron.nextcall, fields.Datetime.now() + timedelta(days=1))
+            self.assertEqual(self.cron.numbercall, -1)
+            self.assertTrue(self.cron.active)
+
+        with self.subTest(doall=False, numbercall=3, missed_call=2), \
+             patch.object(self.registry['ir.cron'], '_callback') as callback:
+            self.cron.active = True
+            self.cron.doall = False
+            self.cron.numbercall = 3
+            self.cron.nextcall = fields.Datetime.now() - timedelta(days=1)
+            process_job()
+            self.assertEqual(callback.call_count, 1)
+            self.assertEqual(self.cron.lastcall, fields.Datetime.now())
+            self.assertEqual(self.cron.nextcall, fields.Datetime.now() + timedelta(days=1))
+            self.assertEqual(self.cron.numbercall, 2)
+            self.assertTrue(self.cron.active)
+
+        with self.subTest(doall=True, numbercall=3, missed_call=2), \
+             patch.object(self.registry['ir.cron'], '_callback') as callback:
+            self.cron.active = True
+            self.cron.doall = True
+            self.cron.numbercall = 3
+            self.cron.nextcall = fields.Datetime.now() - timedelta(days=1)
+            process_job()
+            self.assertEqual(callback.call_count, 2)
+            self.assertEqual(self.cron.lastcall, fields.Datetime.now())
+            self.assertEqual(self.cron.nextcall, fields.Datetime.now() + timedelta(days=1))
+            self.assertEqual(self.cron.numbercall, 1)
+            self.assertTrue(self.cron.active)
+
+        with self.subTest(doall=True, numbercall=3, missed_call=4), \
+             patch.object(self.registry['ir.cron'], '_callback') as callback:
+            self.cron.active = True
+            self.cron.doall = True
+            self.cron.numbercall = 3
+            self.cron.nextcall = fields.Datetime.now() - timedelta(days=3)
+            process_job()
+            self.assertEqual(callback.call_count, 3)
+            self.assertEqual(self.cron.lastcall, fields.Datetime.now())
+            self.assertEqual(self.cron.nextcall, fields.Datetime.now() + timedelta(days=1))
+            self.assertEqual(self.cron.numbercall, 0)
+            self.assertFalse(self.cron.active)
+
+        with self.subTest(trigger=True, numbercall=3, missed_call=0), \
+             patch.object(self.registry['ir.cron'], '_callback') as callback:
+            self.cron.active = True
+            self.cron.numbercall = 3
```

Comment:
there is a lot of duplicate code here. Can you see if you can refactor a little ?